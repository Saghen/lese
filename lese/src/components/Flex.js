import styled from "@emotion/styled";
import Base from "./Base";
import { Children } from "react";

import { propertyGenerator } from "../helpers";

const getFlexProperties = propertyGenerator([
  ["column", () => "flex-direction: column"],
  [
    "xAlign",
    {
      default: "center",
      handler: ({ column, xAlign }) =>
        column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`,
    },
  ],
  [
    "yAlign",
    {
      default: "center",
      handler: ({ column, yAlign }) =>
        column ? `justify-content: ${yAlign}` : `align-items: ${yAlign}`,
    },
  ],
  [
    "noMarginReset",
    () => `> * {
       margin: 0;
      }`,
  ],
  [
    "separation",
    ({ column, separation }) => `
    > * + * {
      ${column ? `margin-top: ${separation}` : `margin-left: ${separation}`}
    }
    `,
  ],

  ["wrap", { default: "wrap", property: "flex-wrap" }],
]);

const getChildFlexProperties = propertyGenerator([
  [
    "xAlignSelf",
    {
      default: "center",
      handler: ({ column, xAlignSelf }) =>
        column ? `align-self: ${xAlignSelf}` : `justify-self: ${xAlignSelf}`,
    },
  ],
  [
    "yAlignSelf",
    {
      default: "center",
      handler: ({ column, yAlignSelf }) =>
        column ? `justify-self: ${yAlignSelf}` : `align-self: ${yAlignSelf}`,
    },
  ],
]);

export default styled(Base)`
  display: flex;
  ${getFlexProperties}
  ${({ children, column }) => {
    const childrenArray = Children.toArray(children).filter(
      (elem) => typeof elem !== "string"
    );

    const properties = [];
    for (const [i, { props }] of childrenArray.entries())
      if (props && (props.xAlignSelf || props.yAlignSelf))
        properties.push(`
        > *:nth-child(${i + 1}) /* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */ {
          ${getChildFlexProperties({ ...props, column })}
        }
        `);
    return properties;
  }}
`;
