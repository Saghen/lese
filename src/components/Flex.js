import { css } from "@emotion/core";
import styled from "@emotion/styled";
import Base from "./Base";

import { propertyGenerator } from "../helpers";
import { CLIENT_RENEG_LIMIT } from "tls";

const getFlexProperties = propertyGenerator([
  ["column", () => "flex-direction: column"],
  [
    "xAlign",
    {
      default: "center",
      handler: ({ column, xAlign }) =>
        column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`
    }
  ],
  [
    "yAlign",
    {
      default: "center",
      handler: ({ column, yAlign }) =>
        column ? `justify-content: ${yAlign}` : `align-items: ${yAlign}`
    }
  ],
  [
    ({ noMarginReset }) =>
      !noMarginReset &&
      `> * {
       margin: 0;
      }`
  ],
  [
    "separation",
    ({ column, separation }) => `
    > * + * {
      ${column ? `margin-top: ${separation}` : `margin-left: ${separation}`}
    }
    `
  ],

  ["wrap", { default: "wrap", property: "flex-wrap" }]
]);

const getChildFlexProperties = propertyGenerator([
  [
    "xAlignSelf",
    {
      default: "center",
      handler: ({ column, xAlignSelf }) =>
        column ? `align-self: ${xAlignSelf}` : `justify-self: ${xAlignSelf}`
    }
  ],
  [
    "yAlignSelf",
    {
      default: "center",
      handler: ({ column, yAlignSelf }) =>
        column ? `justify-self: ${yAlignSelf}` : `align-self: ${yAlignSelf}`
    }
  ]
]);

export default styled(Base)`
  display: flex;
  ${getFlexProperties}
  ${({ children, column }) => {
    if (!Array.isArray(children)) return;
    const properties = [];
    for (const [i, { props }] of children.entries())
      if (props.xAlignSelf || props.yAlignSelf)
        properties.push(`
        > *:nth-child(${i + 1}) {
          ${getChildFlexProperties({ ...props, column })}
        }
        `);
    return properties;
  }}
`;
