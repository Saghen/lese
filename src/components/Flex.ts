import React from "react";
import styled from "@emotion/styled";
import Base from "./Base";
import { propertyGenerator } from "../helpers";

export interface FlexProps {
  column?: boolean;
  reverse?: boolean;
  align?: boolean;
  xAlign?: boolean | string;
  yAlign?: boolean | string;
  separation?: string | string[];
  wrap?: boolean | string;
}

const ignoreFlag =
  "/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */";

const separationHandler = ({ column, separation }: FlexProps) => {
  const isString = typeof separation === "string";
  const isStringArray = Array.isArray(separation);
  if (!isString && !isStringArray) {
    throw new Error("separation must be a string or string array");
  }

  const getStyleBlock = (selector: string, separation: string) => `
      > ${selector} ${ignoreFlag} {
        ${column ? `margin-top: ${separation}` : `margin-left: ${separation}`};
        ${ignoreFlag}
      }
    `;

  const separations = isStringArray ? separation : separation.split(" ").filter(Boolean);
  if (separations.length > 1 || isStringArray) {
    const isRestSeparation = (separation: string) => separation.startsWith("...");
    const restSeparations = separations.filter(isRestSeparation);
    if (restSeparations.length > 1) {
      throw new Error("separation can only contain one rest operator");
    }

    const restSeparationIndex = separations.findIndex((a) => isRestSeparation(a));
    const startingSeparations = separations.slice(
      0,
      restSeparationIndex === -1 ? separations.length : restSeparationIndex
    );
    const endingSeparations = separations.slice(
      restSeparationIndex === -1 ? separations.length : restSeparationIndex + 1
    );

    return [
      ...restSeparations.map((separation) => getStyleBlock("* + *", separation.slice(3))),
      // Order is important since we want startingSeparations to replace endingSeparations
      ...endingSeparations.map((separation, i) =>
        getStyleBlock(`*:nth-last-child(${endingSeparations.length - i})`, separation)
      ),
      ...startingSeparations.map((separation, i) =>
        getStyleBlock(`*:first-child ${new Array(i + 1).fill("+ *").join(" ")}`, separation)
      ),
    ].join("\n");
  }
  return getStyleBlock("* + *", separation);
};

const applyDefaultProps =
  <T>(defaultProps: Partial<T>, callback: (arg: T) => any) =>
  (props: T) =>
    callback({ ...defaultProps, ...props });

const getFlexProperties = propertyGenerator<FlexProps>([
  ["column", () => `flex-direction: column`],
  ["reverse", { handler: ({ column }) => `flex-direction: ${column ? "column" : "row"}-reverse` }],
  [
    "align",
    {
      default: "center",
      handler: ({ align }) => `align-items: ${align}; justify-content: ${align}`,
    },
  ],
  [
    "xAlign",
    {
      default: "center",
      handler: ({ column, xAlign }) => (column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`),
    },
  ],
  [
    "yAlign",
    {
      default: "center",
      handler: ({ column, yAlign }) => (column ? `justify-content: ${yAlign}` : `align-items: ${yAlign}`),
    },
  ],
  ["separation", separationHandler],
  ["wrap", { default: "wrap", property: "flex-wrap" }],
]);

const getRowProperties = applyDefaultProps({ column: false }, getFlexProperties);
const getColumnProperties = applyDefaultProps({ column: true }, getFlexProperties);

export default styled(Base)<React.PropsWithChildren<FlexProps>>`
  display: flex;
  ${getFlexProperties}
`;

export const Row = styled(Base)<React.PropsWithChildren<FlexProps>>`
  display: flex;
  ${getRowProperties}
`;

export const Column = styled(Base)<React.PropsWithChildren<FlexProps>>`
  display: flex;
  ${getColumnProperties}
`;
