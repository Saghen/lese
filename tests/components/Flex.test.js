import React from "react";
import renderer from "react-test-renderer";
import Flex from "../../src/components/Flex";
import { matchers } from "jest-emotion";

expect.extend(matchers);

test("flex props", () => {
  const treeColumn = renderer
    .create(
      <Flex column xAlign="flex-start" yAlign="flex-end" separation="16px">
        <div>Test</div>
      </Flex>
    )
    .toJSON();

  const treeRow = renderer
    .create(
      <Flex xAlign="flex-start" yAlign="flex-end" separation="16px" wrap>
        <div>Test</div>
      </Flex>
    )
    .toJSON();

  expect(treeColumn).toMatchSnapshot();
  expect(treeRow).toMatchSnapshot();
});

test("flex child props", () => {
  const treeColumn = renderer
    .create(
      <Flex column>
        <div xAlignSelf="flex-start" yAlignSelf="flex-end">
          Test
        </div>
      </Flex>
    )
    .toJSON();
  const treeRow = renderer
    .create(
      <Flex>
        <div xAlignSelf="flex-start" yAlignSelf="flex-end">
          Test
        </div>
      </Flex>
    )
    .toJSON();

  expect(treeColumn).toMatchSnapshot();
  expect(treeRow).toMatchSnapshot();
});
