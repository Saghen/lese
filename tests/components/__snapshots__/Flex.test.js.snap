// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flex child props 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 > * {
  margin: 0;
}

<div
  className="emotion-0"
>
  <div
    xAlignSelf="flex-start"
    yAlignSelf="flex-end"
  >
    Test
  </div>
</div>
`;

exports[`flex child props 2`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 > * {
  margin: 0;
}

<div
  className="emotion-0"
>
  <div
    xAlignSelf="flex-start"
    yAlignSelf="flex-end"
  >
    Test
  </div>
</div>
`;

exports[`flex props 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.emotion-0 > * {
  margin: 0;
}

.emotion-0 > * + * {
  margin-top: 16px;
}

<div
  className="emotion-0"
>
  <div>
    Test
  </div>
</div>
`;

exports[`flex props 2`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-0 > * {
  margin: 0;
}

.emotion-0 > * + * {
  margin-left: 16px;
}

<div
  className="emotion-0"
>
  <div>
    Test
  </div>
</div>
`;
